### PR TITLE
Add weight support for eff calculation

### PR DESCRIPTION
**Pull request by dkobylianskii**
## Summary

This pull request introduces the following changes

* Improve the *efficiency* of the efficiency calculation
* Add weight support
* Fix possible input issue with `IntegratedEff` plot 

Relates to the following issues

* #197 

## Conformity
- [x] [Changelog entry](https://github.com/umami-hep/puma/blob/main/changelog.md)
- [x] [Documentation](https://umami-hep.github.io/puma/)

### CODE REVIEW

File: puma/metrics.py [hunk 1]
```diff
@@ -5,11 +5,45 @@
...
+    -------
+    np.ndarray
+        Weighted percentile array
+    """
+    if weights is None:
+        weights = np.ones_like(data)
+    ix = np.argsort(data)
+    data = data[ix]  # sort data
+    weights = weights[ix]  # sort weights
+    cdf = np.cumsum(weights) - 0.5 * weights
```
Comment (samvanstroud) [inline puma/metrics.py:34]: why do you halve the weights here?

File: puma/metrics.py [hunk 2]
```diff
@@ -5,11 +5,45 @@
...
+    cdf -= cdf[0]
```
Comment (samvanstroud) [inline puma/metrics.py:35]: I'm also unclear on why you apply this, might be helpful to add some comments to these lines

**Review by samvanstroud — CHANGES_REQUESTED** [review]
Thanks a lot @dkobylianskii! I had a couple of comments. Pinging also @dguest in case he wants to take a look here. We appear to be missing `inf` under/overflows as [suggested by Dan](https://github.com/umami-hep/puma/issues/197#issuecomment-1660374682).

File: puma/integrated_eff.py [hunk 3]
```diff
@@ -48,8 +48,12 @@ def __init__(
...
         """
         super().__init__(**kwargs)
-        self.disc_sig = disc_sig
-        self.disc_bkg = disc_bkg
+        self.disc_sig = (
+            disc_sig if isinstance(disc_sig, np.ndarray) else np.array(disc_sig)
+        )
+        self.disc_bkg = (
+            disc_bkg if isinstance(disc_bkg, np.ndarray) else np.array(disc_bkg)
+        )
```
[the remark this answers is not in the document]
Comment (dkobylianskii) [reply puma/integrated_eff.py]: Changed

File: puma/metrics.py (hunk 1 above)
Comment (dkobylianskii) [reply puma/metrics.py:39]: For the possibility to weight also signal events. If they have weights, we treat each of them not as `1` but as its corresponding weight and then calculate the percentile. I added links to Wiki and Stackoverflow.

File: puma/metrics.py [hunk 4]
```diff
@@ -37,19 +75,25 @@ def calc_eff(
     """
     # TODO: with python 3.10 using type union operator
     # float | np.ndarray for both target_eff and the returned values
+    if isinstance(target_eff, list):
+        target_eff = np.array(target_eff)
+
+    cutvalue = weighted_percentile(sig_disc, 1.0 - target_eff, weights=sig_weights)
     if isinstance(target_eff, float):
```
[the remark this answers is not in the document]
Comment (dkobylianskii) [reply puma/metrics.py]: Switched to `np.asarray`

File: puma/metrics.py [hunk 5]
```diff
@@ -19,37 +58,48 @@ def calc_eff(
...
+        return_float = True
+
+    target_eff = np.asarray([target_eff]).flatten()
+
+    cutvalue = weighted_percentile(sig_disc, 1.0 - target_eff, weights=sig_weights)
+    sorted_args = np.argsort(
+        1 - target_eff
+    )  # need to sort the cutvalues to get the correct order
+    hist, _ = np.histogram(
+        bkg_disc, (-np.inf, *cutvalue[sorted_args], np.inf), weights=bkg_weights
```
Comment (dkobylianskii) [inline puma/metrics.py:94]: > We appear to be missing `inf` under/overflows as [suggested by Dan](https://github.com/umami-hep/puma/issues/197#issuecomment-1660374682).

I handle them here

Comment (samvanstroud) [reply puma/metrics.py:94]: Thanks

File: puma/metrics.py (hunk 1 above)
Comment (samvanstroud) [reply puma/metrics.py:39]: Thanks for the links!

**Comment by samvanstroud** [comment]
This is looking great. Can you add weights to the ROC class too in order to complete the implementation?

**Review by samvanstroud — COMMENTED** [review]
Looks good to me! SHould

**Review by samvanstroud — APPROVED** [review]

**Comment by dguest** [comment]
Minor commnets:
- The function has a help string that says "calculates efficiency", but it calculates the _background_ efficiency for a given signal efficiency target, right? (not really a problem, I just had to read the code to understand what was going on)
- This implementation might be better than the one I imagined 😀. You loop over the data twice, once with the fancy wikipedia thing to get the cut values and again to get the background efficiency at those values. I was imagining that you'd make a histogram and then use the bin edges and efficiencies to interpolate the cut values. I think both might have advantages: 
  - Your way is certainly more exact
  - It might also be slower given all the sorting it requires
  - It also doesn't trivially extend to huge datasets that don't fit in memory (but given that the function takes an array anyway reworking the internals can't solve this)

Anyway unless people are regularly running out of memory this seems like a big improvement: hopefully at least making a ROC with 1000 cut values is much faster with this.

**Comment by samvanstroud** [comment]
As far as I know no one has had memory issues with jet-wise stuff yet :D

**Comment by dkobylianskii** [comment]
Hi @dguest, @samvanstroud ,

I checked the speed, we have 100 times improvement for 1e6 jets and 1e3 cuts :)
In the ROC class, we use recalculated rejection values, so there is no need to add weights there.

**Comment by samvanstroud** [comment]
@dkobylianskii thanks!

**Pull request merged**